engine/src/main/java/org/pentaho/di/www/ExecuteTransServlet.java

```diff
@@ -434,4 +336,214 @@ public String getContextPath() {
+      if ( isPostCall ) {
+        previewDataMap = executePostTrans( trans );
+      } else {
+        executeTrans( trans );
+      }
+
+      String logging = KettleLogStore.getAppender().getBuffer( trans.getLogChannelId(), false ).toString();
+      if ( trans.isFinishedOrStopped() && trans.getErrors() > 0 ) {
+        response.setStatus( HttpServletResponse.SC_INTERNAL_SERVER_ERROR );
+        out.println( new WebResult( WebResult.STRING_ERROR, BaseMessages.getString(
+                PKG, "ExecuteTransServlet.Error.ErrorExecutingTrans", logging ) ) );
+      }
+      if ( isPostCall ) {
+        writeToJson( out, transMeta, carteObjectId, previewDataMap );
+      }
+      out.flush();
+    } catch ( Exception executionException ) {
+      String logging = KettleLogStore.getAppender().getBuffer( trans.getLogChannelId(), false ).toString();
+      throw new KettleException( BaseMessages.getString( PKG, "ExecuteTransServlet.Error.ErrorExecutingTrans", logging ), executionException );
+    }
+  }
+
+
+  private Map<StepMeta, List<RowMetaAndData>> executePostTrans( Trans trans ) throws KettleException {
+    trans.prepareExecution( null );
+    Map<StepMeta, List<RowMetaAndData>> previewDataMapUpdated = null;
+    previewDataMapUpdated = capturePreviewData( trans, trans.getTransMeta().getSteps(), new HashMap<>() );
+    trans.startThreads();
+    trans.waitUntilFinished();
+    return previewDataMapUpdated;
+  }
+
+  private void writeToJson( PrintWriter out, TransMeta transMeta, String carteObjectId, Map<StepMeta, List<RowMetaAndData>> previewDataMap ) {
+    JSONObject finalJsonOutput = new JSONObject();
+    finalJsonOutput.put( "carteId", carteObjectId );
+
+    JSONArray previewJson = new JSONArray();
+
+    for ( StepMeta stepMeta : transMeta.getSteps() ) {
+
+      JSONObject stepJSON = new JSONObject();
+      stepJSON.put( "stepName", stepMeta.getName() );
+
+      List<RowMetaAndData> rowMetaAndDataList = previewDataMap.get( stepMeta );
+      String[] columnNames = rowMetaAndDataList.get( 0 ).getRowMeta().getFieldNames();
+
+      stepJSON.put( "columnInfo", Arrays.asList( columnNames ) );
+      JSONArray dataArray = new JSONArray();
+      String data = "";
+      try {
+        for ( RowMetaAndData rowMetaAndData : rowMetaAndDataList ) {
+          Object[] rowData = rowMetaAndData.getData();
+          RowMetaInterface rowMeta = rowMetaAndData.getRowMeta();
+          JSONArray dataRowArray = new JSONArray();
+
+          for ( int column = 0; column < columnNames.length; column++ ) {
+            data = rowMeta.getString( rowData, column );
+            dataRowArray.add( data );
+          }
+          JSONObject dataRow = new JSONObject();
+          dataRow.put( "data", dataRowArray );
+          dataArray.add( dataRow );
+        }
+        stepJSON.put( "rows", dataArray );
+      }  catch ( Exception e ) {
+        data = "Conversion error: " + e.getMessage();
```

Comment:
when exception happens multiple times, data will contain only the last time message... and this data is used somewhere for logging ?

Reply:
This data will be used in the future. This is being handled in Backlog-38964